avatar/metrics/trace.py

```diff
@@ -0,0 +1,241 @@
+            ),
+            trusted_packet_sequence_id=self.device.process_id,
+        )
+
+
+class CallEvent(AsTrace):
+    def __init__(self, callsite: Callsite, message: Any) -> None:
+        self.at = time.monotonic_ns() - genesis
+        self.callsite = callsite
+        self.message = message
+
+        callsite.device.log.info(f"{self}")
+
+    def __str__(self) -> str:
+        return "└── " + self.stringify('->')
+
+    def as_trace(self) -> TracePacket:
+        return TracePacket(
+            timestamp=self.at,
+            track_event=TrackEvent(
+                name=self.callsite.name,
+                type=TrackEvent.Type.TYPE_INSTANT,
+                track_uuid=self.callsite.device.uuid,
+                debug_annotations=None
+                if self.message is None
+                else [DebugAnnotation(name=self.message.__name__, string_value=message_prettifier(f"{self.message}"))],
+            ),
+            trusted_packet_sequence_id=self.callsite.device.process_id,
+        )
+
+    def stringify(self, direction: str) -> str:
+        message = "" if self.message is None else self.message
+        message_pretty = message_prettifier(f"{message}")
+        return f"[{(self.at - self.callsite.at) / 1000000000:.3f}s] {self.callsite} {direction} ({message_pretty})"
+
+
+class CallOutput(CallEvent):
+    def __str__(self) -> str:
+        return "├── " + self.stringify('->')
+
+    def as_trace(self) -> TracePacket:
+        return super().as_trace()
+
+
+class CallInput(CallEvent):
+    def __str__(self) -> str:
+        return "├── " + self.stringify('<-')
+
+    def as_trace(self) -> TracePacket:
+        return super().as_trace()
+
+
+class CallEnd(CallEvent):
+    def __str__(self) -> str:
+        return "└── " + self.stringify('->')
+
+    def as_trace(self) -> TracePacket:
+        return TracePacket(
+            timestamp=self.at,
+            track_event=TrackEvent(
+                name=self.callsite.name,
+                type=TrackEvent.Type.TYPE_SLICE_END,
+                track_uuid=self.callsite.device.uuid,
+                debug_annotations=None
+                if self.message is None
+                else [DebugAnnotation(name=self.message.__name__, string_value=message_prettifier(f"{self.message}"))],
+            ),
+            trusted_packet_sequence_id=self.callsite.device.process_id,
+        )
+
+
+def message_prettifier(msg: str) -> str:
+    msg = msg.replace("{\n", "{").replace("\n", ",").replace(" ", "").replace(",}", "}")
+    msg = re.sub(r"{cookie{value:(\"[^\"]+\")}}", lambda m: f":{m.groups()[0]}", msg)
+
+    def repl(addr: re.Match[str]) -> str:
+        m = CUnescape(addr.groups()[0])
+        if len(m) > 8:
+            return ":'...'"
+        if len(m) == 4 and m[0] == 0 and m[1] == 0:
+            m = m[2:]
+        return ":'" + ':'.join([f'{x:02X}' for x in m]) + "'"
+
+    return re.sub(r":\"([^\"]+)\"", repl, msg)[:-1]
```

Comment:
This is kinda ugly could we find a better way to do this ?

Reply:
@uael

Reply:
I think it's fine for now, it's only used to logging, a next step would be to recursively generate the debug annotations from the message, but it can be done later (or not at all)

Reply:
I prefer that we either remove it or fix it. By keeping it like that, there is a high chance that it'll stay like that.

Reply:
Done!